packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
   }
 }
 
+//Given array containing entries per row, produce rowptrs
+//Used by all add kernels
+template<typename RowPtrType>
+struct PrefixSumFunctor
+{
+  PrefixSumFunctor(RowPtrType& Crowcounts_, RowPtrType& Crowptrs_) : Crowcounts(Crowcounts_), Crowptrs(Crowptrs_) {}
+  //note: call this functor over the full length of Crowptrs (nrows+1)
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i, size_t& update, const bool finalPass) const
+  {
+    // exclusive scan
+    if(finalPass)
+    {
+      Crowptrs(i) = update;
+    }
+    update += Crowcounts(i);
+    if(finalPass)
+    {
+      if(i == Crowcounts.dimension_0() - 1)
+      {
+        Crowptrs(i + 1) = update;
+      }
+    }
+  }
+  RowPtrType& Crowcounts;
+  RowPtrType& Crowptrs;
+};
+
+/****************************************************
+* Fast local thread-parallel addition using Kokkos *
+ * Adds two Kokkos local CRS matrices               *
+ * Two main code paths: A,B sorted or nonsorted     *
+ ****************************************************/
+namespace AddSortedFunctors
+{
+
+template<typename Ordinal, typename RowPtrs, typename ColInds>
+struct EntryCountingFunctor
```

Comment:
@mhoemmen Combined the entry count and prefix sum is about twice as slow as doing them separately. I think that's because the expensive part (scanning entries in rows and matching indices) is being executed on every non-final pass of scan. Before I was caching all the entry counts first and then prefix summing. I think I'll have to leave them separate.